fabtests/functional/rdm_stress.c

```diff
@@ -0,0 +1,479 @@
+
+	size = sizeof(*resp) + req->size;
+	resp = calloc(1, size);
+	if (!resp)
+		return -FI_ENOMEM;
+
+	ret = rpc_inject(req);
+	if (ret)
+		goto free;
+
+	ret = rpc_trecv(resp, size, req->data);
+	if (ret)
+		goto close;
+
+	assert(resp->cmd == cmd_tag);
+	ret = ft_check_buf(resp + 1, req->size);
+
+close:
+	fi_close(&mr->fid);
+free:
+	free(resp);
+	return ret;
+}
+
+/* Request server to perform an RMA read from our buffer */
+static int rpc_read(struct rpc_hdr *req)
+{
+	struct rpc_hdr resp;
+	char *buf;
+	size_t size;
+	int ret;
+
+	size = req->offset + req->size;
+	buf = calloc(1, size);
+	if (!buf)
+		return -FI_ENOMEM;
+
+	ft_fill_buf(&buf[req->offset], req->size);
+	ret = fi_mr_reg(domain, buf, size, FI_REMOTE_READ, 0, rpc_read_key,
+			0, &mr, NULL);
+	if (ret)
+		goto free;
+
+	req->data = fi_mr_key(mr);
+	ret = rpc_inject(req);
+	if (ret)
+		goto close;
+
+	ret = rpc_recv(&resp, sizeof(resp));
+	if (ret)
+		goto close;
+
+	assert(resp.cmd == cmd_read);
+	ret = ft_check_buf(&buf[req->offset], req->size);
+
+close:
+	fi_close(&mr->fid);
+free:
+	free(buf);
+	return ret;
+}
+
+/* Request server to perform an RMA write to our buffer */
+static int rpc_write(struct rpc_hdr *req)
+{
+	struct rpc_hdr resp;
+	char *buf;
+	size_t size;
+	int ret;
+
+	size = req->offset + req->size;
+	buf = calloc(1, size);
+	if (!buf)
+		return -FI_ENOMEM;
+
+	ret = fi_mr_reg(domain, buf, size, FI_REMOTE_WRITE, 0, rpc_write_key,
+			0, &mr, NULL);
+	if (ret)
+		goto free;
+
+	req->data = fi_mr_key(mr);
+	ret = rpc_inject(req);
+	if (ret)
+		goto close;
+
+	ret = rpc_recv(&resp, sizeof(resp));
+	if (ret)
+		goto close;
+
+	assert(resp.cmd == cmd_write);
+	ret = ft_check_buf(&buf[req->offset], req->size);
+
+close:
+	fi_close(&mr->fid);
+free:
+	free(buf);
+	return ret;
+}
+
+/* Used to delay client, which can force server into a flow control
+ * state or into the middle of a transfer when the client exits.
+ */
+static int rpc_sleep(struct rpc_hdr *hdr)
+{
+	int ret;
+
+	ret = usleep((useconds_t) hdr->size * 1000);
+	return ret ? -errno : 0;
+}
+
+static int rpc_exit(struct rpc_hdr *hdr)
+{
+	exit(0);
+	return -1;
+}
+
+
+int (*rpc_req[cmd_last])(struct rpc_hdr *req) = {
+	rpc_hello,
+	rpc_goodbye,
+	rpc_msg,
+	rpc_tag,
+	rpc_read,
+	rpc_write,
+	rpc_sleep,
+	rpc_exit,
+};
+
+static int run_child(void)
+{
+	int i, ret;
+
+	ret = ft_init_fabric();
+	if (ret)
+		return ret;
+
+	for (i = 0; i < rpc_cnt; i++) {
+		if (rpc[i].client_id != myid)
+			continue;
+
+		ret = rpc_req[rpc[i].cmd](&rpc[i]);
+		if (ret)
+			break;
+	}
+
+	ft_free_res();
+	return ret;
+}
+
+static int init_rpc(void)
+{
+	/* TODO: read and parse control file.
+	 * Will allocate and format rpc array.
+	 */
+	return -ENOSYS;
+}
+
+static int run_parent(void)
+{
+	int i, ret;
+
+	ret = init_rpc();
+	if (ret)
+		return ret;
+
+	for (i = 0; i < opts.iterations; i++) {
+		for (myid = 0; myid < opts.num_connections; myid++) {
+			if (clients[myid].pid) {
+				waitpid(clients[myid].pid, NULL, 0);
```

Comment:
Would it be better to use wait() or waitpid(-1, ..) to catch any child process instead wait for them in turn?